Correct password option docs and add related tests

fidgetingbits opened the pull request:
## Description of changes

This attempts to clarify the true behavior of password option overrides, as the documentation seems to be wrong (and appears to have been for at least 10 years). These docs also clarify systemd-sysusers quirks that I coincidentally noticed someone happened to file recently [here](https://github.com/NixOS/nixpkgs/issues/307159) as well.

The relevant option docs have been updated, but more importantly I created two new tests to try to show the true behavior of the overrides in their current form. This way if anything ever changes on this front it will be caught and docs can be updated accordingly.

If you want to run the tests I added, you will have to change the maintainer for now as my being added to maintainers is pending [this PR](https://github.com/NixOS/nixpkgs/pull/296562) being merged.

I've not used Nix for very long, so would appreciate someone who has dealt with this code in the past to validate what I've found. @nikstur and @NeQuissimus, perhaps you could both take a look as I've repurposed some of your tests that touch the same areas, so I suspect you'll both be familiar. 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3958

fidgetingbits commented:
@Ma27 any thoughts on this PR or know who should review it? I just noticed today you recently made some recent changes that lead to a merge conflict (which I have yet to fix), so just pinging to see if you have want to take a look or have an idea who to ping to get this reviewed.

Ma27 commented:
Last times I touched this part, it was kinda hard to find anybody who felt responsible.
Given that this seems to be a correctness fix mostly, I'll try to take a look soonish.

fidgetingbits commented:
Merge conflicts should be fixed now

Ma27 commented on nixos/modules/config/users-groups.nix:
```diff
@@ -352,7 +357,8 @@ let
           Note that the {option}`password` option will override this
-          option if both are set.
+          option if both are set. One exception to this rule is if
+          `systemd.sysusers.enable = true;`.
```
Hmm after reading through the code it seems to me as if all non-initial password options would override this.

fidgetingbits replied:
I might not understand what you mean, but I think the test on line 35 of nixos/tests/password-option-override-ordering.nix that I added shows that initialPassword actually overrides hashedPassword (a non-initial password option), which conflicts with what you are saying?

Ma27 replied:
Only for `mutableUsers = false`; though, see https://github.com/NixOS/nixpkgs/blob/400e3ecf59205c6197e093e571e6518a669c4d55/nixos/modules/config/users-groups.nix#L417-L421

Ma27 commented on nixos/modules/config/users-groups.nix:
```diff
@@ -69,6 +69,10 @@ let
     user passwords, they will always be set according to the password
     options.
+
+    NOTE: one exception is if `systemd.sysusers.enable = true;`. In this case
```
I think you can cross-reference this option with `[](#opt-systemd.sysusers.enable)`.

Ma27 commented on nixos/modules/config/users-groups.nix:
```diff
@@ -960,11 +966,14 @@ in {
         set at most one of the options above to a non-`null` value.
+        NOTE: If `systemd.sysusers.enable = true;` then the the options and order
+        of precedence are: `initialPassword`,  `initialHashedPassword`,
+        & `hashedPasswordFile`.
```
What about the other two options?

fidgetingbits replied:
The other two options aren't supported when using `systemd.sysusers.enable`. I changed that to use some other (hopefully more clear) wording from elsewhere. So should be addressed, but let me know if you still think it's too unclear.

Ma27 commented on nixos/modules/config/users-groups.nix:
```diff
@@ -960,11 +966,14 @@ in {
         The user '${user.name}' has multiple of the options
-        `hashedPassword`, `password`, `hashedPasswordFile`, `initialPassword`
-        & `initialHashedPassword` set to a non-null value.
+        `initialHashedPassword`, `hashedPassword`, `initialPassword`, `password`,
```
The trailing comma should go here: there's an `&` the next char.

fidgetingbits replied:
TIL

fidgetingbits commented:
Just a note that to run the tests I added you can run the following from nixpkgs root folder: `nix-build . -A nixosTests.password-option-override-ordering` and `nix-build . -A nixosTests.systemd-sysusers-password-option-override-ordering`. 

Looks like the latter is failing now due to a new assertion about user types, so I will have to investigate.

fidgetingbits commented:
Looks like there are more aggressive asserts that helps with some of the systemd-sysusers stuff, so I removed a couple of assertions from that test that no longer will run. Also changed the user type and added groups to ensure other new assertions pass so the test actually runs.

I've left the changes as separate commit now for sake of reference, but can squash them before merge.

Ma27 commented on nixos/tests/password-option-override-ordering.nix:
```diff
@@ -0,0 +1,151 @@
+          machine.send_chars("egon\n")
+          machine.wait_until_tty_matches("2", "login: egon")
+          machine.wait_until_succeeds("pgrep login")
+          machine.sleep(2)
```
This seems a little brittle as it only depends on the timing. After reading through all the checks: is that really necessary?

Ma27 replied:
Oof.

But well, shadow is part of the tested set, so it seems to be surprisingly stable :shrug:

fidgetingbits replied:
Actually I spent some time digging around / testing while I was deduplicating the code and it turns out it's easy to drop the sleep. I'll plan to send some PRs to drop the sleeps from `shadow.nix` and any others I find to speed up tests in the future.

Ma27 commented on nixos/tests/password-option-override-ordering.nix:
```diff
@@ -0,0 +1,151 @@
+
+      with subtest("greg user has correct password"):
+        print(machine.succeed("getent passwd greg"))
+        assert "${hashed_sha512crypt}" not in machine.succeed("getent shadow greg"), "greg user password is not correct"
```
We could match the output of the hashed password from `getent` against the password set for greg here I think.

Also, it's a little weird that you return `passwd` rather than shadow above?

Ma27 replied:
> Can only check exact match when providing the hash yourself.

Sure, but there are tools that you can use to verify the hash against the plaintext password.

> By return you mean what is printed ya? I switched them all to print getent shadow. Makes sense.

Correct.

fidgetingbits replied:
> > Can only check exact match when providing the hash yourself.
> 
> Sure, but there are tools that you can use to verify the hash against the plaintext password.

Ok. I'll add that

Ma27 commented on nixos/modules/config/users-groups.nix:
```diff
@@ -57,9 +57,9 @@ let
-    {option}`password` overrides {option}`hashedPasswordFile`.
+    {option}`hashedPasswordFile` overrides both
+    {option}`password` and {option}`hashedPassword`.
+    {option}`hashedPassword` overrides {option}`password`.
```
Line 57 of the password-option-* test says `hashedPassword -> ... -> password`, so the override is the other way round here, right?

Ma27 replied:
Also given that this also depends on `mutableUsers` somehow, I guess this is also worth mentioning.

Ma27 reviewed (changes requested):
Oof thank you for investing time into this! Using NixOS for almost 8 years now and I had no idea _how_ much of a mess that is...

Ma27 commented:
Also asked in the nixos-systemd Matrix channel for more reviewers for the systemd-sysusers part.

Ma27 commented:
I'd really like to see this in 24.11 when it's good to go: this isn't a behavioral fix, but clarifies the current behavior in both warnings and the manual (and locks it down with a test).

Ma27 commented on nixos/tests/password-option-mutable-override-ordering.nix:
```diff
@@ -0,0 +1,150 @@
+        environment.systemPackages = [ pkgs.shadow ];
+
+        users = {
+          mutableUsers = true;
```
Wouldn't it make sense to merge this with the immutable test, i.e. have two nodes in there and deduplicate the config and assertions as much as possible?

fidgetingbits replied:
Done, also deduplicated the login test part.

fidgetingbits commented:
I still need to add the clearer wording for the greg case with mutableUsers difference, but otherwise all other feedback should be addressed now.

EmergentMind commented on nixos/modules/config/users-groups.nix:
```diff
@@ -329,8 +333,10 @@ let
+          Note that, when {option}`users.mutableUsers` is true, the {option}`hashedPassword`
+          option will override both the {option}`user.initialHashedPassword` option and
+          `{option}`user.initialPassword` if set. One exception to this rule is if
+          [](#opt-systemd.sysusers.enable) is true.
```
What is the exception?
Perhaps editing to something like: "... is true, in which case only `initialPassword`, `initialHashedPassword`, and `hashedPasswordFile` options will be used."

This also applies to Note one line 360

EmergentMind reviewed (changes requested):
Some suggestions to add clarity and improve some grammar. 

I also reviewed the various cases of precedence depending on the values of mutable and/or sysUsers. All the option descriptions and related tests appear to be consistent.

fidgetingbits commented:
I took a stab at rewording the whole thing based on EmergentMind's feedback and because I found myself getting confused reading it even though I'm working on it. That allowed me to duplicate some text from other options, which makes `passwordDescription` the single source of truth, which will be nice for future tweaks. 

One part I didn't include which I considered putting at the end of `passwordDescription` is:

```
    If [](#opt-systemd.sysusers.enable) is `true`, then the order of password is:
    `initialPassword` -> `initialHashedPassword` -> `hashedPasswordFile`
```

Since technically that order still matches what is shown earlier (just with password, hashedPassword, hashedPasswordFile omitted), I wasn't sure if I should explicitly include it or not. Happy to add it if others thing it's worth.

fidgetingbits commented on nixos/modules/config/users-groups.nix:
```diff
@@ -960,12 +979,16 @@ in {
-        & `initialHashedPassword` set to a non-null value.
+        `initialHashedPassword`, `hashedPassword`, `initialPassword`, `password`
+        & `hashedPasswordFile` set to a non-null value.
         The options silently discard others by the order of precedence
```
I guess this is not true because of the mutableUsers true/false difference, so the order of precedence isn't necessary as shown above. Would be nice to refer to a non-duplicated source of truth here as well, so maybe I'll add a separate variable with the order for mutable false/true text and then include that both in passwordDescription and here.. 

I also wonder if the order of precedence difference between mutableUsers true/false is intentional or accidental. If the former, I wonder who would know, and if the latter, I think it should just be changed for a future release so the orders are consistent?

fidgetingbits commented:
Good for another review pass I think. In order to fix the assert warning being wrong, and make it harder to mess up the override order text in various places, I tried to deduplicate it all and use some variable expansion to help generate it in various places.

EmergentMind reviewed (commented):
The rewording and de-duplication is a big improvement for clarity. Nice work.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4905

Ma27 commented on nixos/modules/config/users-groups.nix:
```diff
@@ -53,37 +53,80 @@ let
 
+  overrideOrderMutable = ''
+    {option}`initialHashedPassword` -> {option}`initialPassword` ->
+    {option}`hashedPassword` -> {option}`password`
```
Isn't there an arrow missing? Same for the ordering below.

fidgetingbits replied:
Thanks for catching that!

Ma27 reviewed (commented):
Code LGTM otherwise.
Please reword your commit messages to follow the contribution guidelines.

fidgetingbits commented on nixos/modules/config/users-groups.nix:
```diff
@@ -53,37 +53,80 @@ let
+    `${builtins.toString isMutable}`, then the order of precedence is as shown
+    below, where values on the left are overridden by values on the right:
+
+    ${if isMutable then overrideOrderMutable else overrideOrderImmutable}
```
I've rebased with new commits following the contribution guideline and at the same time I _think_ I addressed this suggestion, but please take a look at confirm it's what you wanted. 

I will look into how to actually generate the new docs from nixpkgs, so I can confirm things are formatted sanely.

fidgetingbits replied:
Hah.. there is indeed a rendering error I'll fix, but figured I'd note it down here anyway. Built with `nix-build nixos/release.nix -A manual.x86_64-linux` and checked the `options.html`:

> If the option users.mutableUsers is ``, then the order of precedence is as shown below, where values on the left are overridden by values on the right: initialHashedPassword -> hashedPassword -> initialPassword -> password -> hashedPasswordFile

> If the option users.mutableUsers is 1, then the order of precedence is as shown below, where values on the left are overridden by values on the right: initialHashedPassword -> initialPassword -> hashedPassword -> password -> hashedPasswordFile


Is because:
```
nix-repl> builtins.toString true
"1"

nix-repl> builtins.toString false
""
```

fidgetingbits replied:
The rendering issue is fixed now

Ma27 replied:
> Is because:

Heh, didn't even notice it during review.

Fwiw, this is (IIRC) because Nix does `toString` on all derivation attributes before putting those into the env to make them accessible for the builder (almost always sh/bash) and it kinda makes sense to have an empty string for false in such a language. But yeah, it's a surprise the first time for everybody ;-)

fidgetingbits replied:
Thanks for the insight! It seemed a bit cursed at first, but in the context of generating shellscript output it does indeed make a lot of sense.

fidgetingbits committed:
nixos/users-groups: Correct and refactor password override documentation

Testing showed that the existing documentation regarding password override
ordering was incorrect. This commit corrects the errors and refactors
the way the text is constructed to make updating future ordering
changes significantly easier.

fidgetingbits committed:
nixos/tests: Add two new tests for password option override ordering

This commit adds two new tests to show that the ordering of password
overrides documentation in nixos/modules/config/user-groups.nix is
correct. The override behavior differs depending on whether a system
has systemd-sysusers enabled, so there are two tests.

Ma27 committed:
nixos/modules: remove a few whitespaces from the pw override parts

Otherwise the evaluation warnings have a two or even three lines of
whitespace between paragraphs.

Ma27 committed:
nixos/tests/systemd-sysusers-password-option-override-ordering: clarify SetCredential check

It was kinda weird to assert that the clear-text password was
in the unit when the hashed password was the effective one.

This change makes it explicit that both are in there and the latter
takes precedence.

Ma27 reviewed (approved):
Pushed two small fixes (one for the whitespace, one for an assertion in the sysusers test - details are in the messages). LGTM now.

fidgetingbits commented:
Thanks a lot for all your help working through this!
